### DISCUSSION TITLE
Removing tiny bodies from a larger shape

### DISCUSSION THREAD

**Discussion by kriho** [sgenoud/replicad#147] [Q&A] [answered]
Hi everyone,

has anyone found a way to detect and remove tiny bodies from a larger shape?
I am generating a model from user input and in some cases small "chips" of a model are left floating around. These bodies all have very short edges. I need a method of finding and removing the "chips" from the main shape.

If anyone found a way or has an idea on how to do this, please let me know! Thanks!

**Comment by raydeleu** [comment]
Are these edges and pieces inside another shape? I think that opencascade.js has an additional function to remove internal edges. The application CascadeStudio, which is also based on opencascade.js, has a function called RemoveInternalEdges(shape, keepShape?). If you look into the code of OpenCascade you can probably derive which OpenCascade function is used to perform this functionality. As far as I know this function is not exposed in Replicad as normally these remnants should be removed by the modelling function automatically.

**Reply by kriho** [reply]
The bodies are floating outside of the main shape. The are all fully closed.

**Reply by sgenoud** [reply]
Is your model a mesh that you then import within replicad?

**Reply by kriho** [reply]
Everything is generated with replicad.
Thanks by the way - I was working with OpenSCAD before and I truly enjoy working with replicad now! I really like the interface and had very few issues so far!

**Reply by sgenoud** [reply, marked as answer]
My guess is that you do a bunch of boolean operations that leave more than one shape, right?

Then you might try to iterate over all the shapes of type body within your compound shape and remove the one you don't want. There are no exposed APIs to list of the "solids" of a generic shape - you could hack by getting inspiration [from here](https://github.com/sgenoud/replicad/blob/c2e7e2c/packages/replicad/src/shapes.ts#L298C17-L298C26)

**Reply by kriho** [reply]
Correct. Thanks for the help, I got it working with your hint.
```
const subShapes = shape._listTopo("solid").map(s => new rep.Solid(s));
for(const subShape of subShapes) {
  const bounds = subShape.boundingBox;
  if(bounds.width < threshold || bounds.depth < threshold) {
    shape= shape.cut(subShape);
  }
}
```
This can lead to a lot of additional boolean operations. Is there a way to just "remove" the filtered sub shapes from the shape?

**Reply by msngupta** [reply]
I wonder if compounding the shapes to be cut using `makeCompound`, then performing a single cut operation on the compound, would be faster?

**Reply by kriho** [reply]
As I understand it, I already have a compound shape and the goal is to remove some of of its shapes. The snippet above works by cutting (boolean operation) the compound shape with one of its shapes. That feels wrong - not sure if that case is detected under the hood or if the boolean operation is performed. I have no solid data but it feels slower now.

**Reply by sgenoud** [reply]
You could probably try something like that:

```js
const subShapes = shape._listTopo("solid").map(s => new rep.Solid(s)).filter(subShape => {
  const bounds = subShape.boundingBox;
  return bounds.width >= threshold && bounds.depth >= threshold
})
```

You would then have a list of shapes you can compound together (or not depending on your use case)

**Reply by kriho** [reply]
That is the easy part, but do I really have to CUT these shapes from the compound shape they are a part of? Can't I just remove them from the compound shape?

**Reply by sgenoud** [reply]
I have inversed the logic - I only select the ones you want to keep - then you can just use them (and ignore the other ones). You build a new compound out of the solids you want and ignore the other ones. Check my assumptions but

```js
const newShape = makeCompound(shape._listTopo("solid").map(s => new rep.Solid(s)).filter(subShape => {
  const bounds = subShape.boundingBox;
  return bounds.width >= threshold && bounds.depth >= threshold
}))```

should just work ™️

**Reply by kriho** [reply]
My specific question about finding tiny solids within a shape is solved. I will create a new question for posterity.

**Reply by kriho** [reply]
Here is the result in the wild: ![signal-2024-04-04-23-57-41-261](https://github.com/sgenoud/replicad/assets/19213657/92541aa6-0f37-4e17-9c8f-71f3e246cd1f)

**Reply by msngupta** [reply]
@kriho  Beautiful UI you have made. Do you use any specific libraries? If your project is open source, would love to learn from it. Thanks

**Reply by kriho** [reply]
I'm not sure what to do with it yet. Check out https://ui.shadcn.com/ and https://tailwindcss.com/
